Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs

```diff
@@ -12,11 +13,49 @@ namespace Microsoft.StreamProcessing.Provider
     /// </summary>
     public abstract class QStreamableVisitor : ExpressionVisitor
     {
-        private static readonly MethodInfo AlterDurationMethod = typeof(QStreamableStatic).GetMethod("AlterDuration");
-        private static readonly MethodInfo GroupByMethod = typeof(QStreamableStatic).GetMethod("GroupBy");
-        private static readonly MethodInfo QuantizeLifetimeMethod = typeof(QStreamableStatic).GetMethod("QuantizeLifetime");
-        private static readonly MethodInfo SelectMethod = typeof(QStreamableStatic).GetMethod("Select");
-        private static readonly MethodInfo WhereMethod = typeof(QStreamableStatic).GetMethod("Where");
+        private static readonly Expression<Func<IQStreamable<object>, IQStreamable<object>>> ChopExpression = s => s.Chop(0, 0);
+        private static readonly MethodInfo ChopMethod = ((MethodCallExpression)ChopExpression.Body).Method.GetGenericMethodDefinition();
```

Comment:
Are the "Expression<Func<...>>" members needed only to construct the MethodInfo members? If so, maybe a static constructor so we don't have to keep the former around as a member?

Reply:
I could go either way on that, though StyleCop at one point frowned on a static constructor for some reason.

Reply:
Interesting, I wonder why

---
In reply to: [309311070](https://github.com/microsoft/Trill/pull/103#discussion_r309311070) [](ancestors = 309311070)

Reply:
Was it SA1642?
If the constructor is static, the summary text should begin with "Initializes static members of the {class name} class." For example:

/// <summary>
/// Initializes static members of the <see cref="Customer"/> class.
/// </summary>
public static Customer()
{
}

---
In reply to: [309311333](https://github.com/microsoft/Trill/pull/103#discussion_r309311333) [](ancestors = 309311333,309311070)

Reply:
No, the one I saw at the time said don't use them. That said, I think that was the old FxCop framework so I'm not sure.

Reply:
CA1810 - learned something new

---
In reply to: [309312580](https://github.com/microsoft/Trill/pull/103#discussion_r309312580) [](ancestors = 309312580)

Reply:
How about this?

        private static readonly MethodInfo ChopMethod = ConvertExpressionToMethodInfo(s => s.Chop(0, 0));
        private static readonly MethodInfo ClipDurationMethod = ConvertExpressionToMethodInfo(s => s.ClipDuration(0));
...
        private static MethodInfo ConvertExpressionToMethodInfo(Expression<Func<IQStreamable<object>, IQStreamable<object>>> expression)
            => ((MethodCallExpression)expression.Body).Method.GetGenericMethodDefinition();

---
In reply to: [309315170](https://github.com/microsoft/Trill/pull/103#discussion_r309315170) [](ancestors = 309315170,309312580)

Reply:
I think that could work - I'll try it out in a supplement.